From 43e56841c28819daed7278585544ea87f625c42d Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 15 Mar 2024 15:38:12 +0100
Subject: [PATCH 0163/1010] net: fec: add oob I/O capabilities

Adapt the FEC driver to the newly introduced oob I/O capabilities of
the netstack. In a nutshell, this entails the following steps:

- turn on/off oob delivery for the NIC interrupts upon request from
  the in-band stack.

- use an oob-accessible page pool for RX.

- make sure TX only synchronizes the CPU caches instead of streaming
  DMA buffers when receiving data living in an oob-accessible page
  pool.

- serialize callers of fec_enet_start_xmit() which run from opposite
  execution stages (i.e. in-band vs oob), so that mixed use is safely
  handled.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 drivers/net/ethernet/freescale/Kconfig    |   9 ++
 drivers/net/ethernet/freescale/fec.h      |  30 +++++
 drivers/net/ethernet/freescale/fec_main.c | 156 +++++++++++++++++++---
 3 files changed, 173 insertions(+), 22 deletions(-)

diff --git a/drivers/net/ethernet/freescale/Kconfig b/drivers/net/ethernet/freescale/Kconfig
index ce866ae3df..0e3414588a 100644
--- a/drivers/net/ethernet/freescale/Kconfig
+++ b/drivers/net/ethernet/freescale/Kconfig
@@ -34,6 +34,15 @@ config FEC
 	  Say Y here if you want to use the built-in 10/100 Fast ethernet
 	  controller on some Motorola ColdFire and Freescale i.MX/S32 processors.
 
+config FEC_OOB
+	bool "Out-of-band support for FEC"
+	depends on NET_OOB
+	help
+	  Enable out-of-band I/O. Say Y if a companion kernel may divert
+	  the network traffic going through FEC devices for real-time delivery.
+	  Enabling this option makes sense only in a dual kernel configuration.
+	  If unsure, disable it.
+
 config FEC_MPC52xx
 	tristate "FEC MPC52xx driver"
 	depends on PPC_MPC52xx && PPC_BESTCOMM
diff --git a/drivers/net/ethernet/freescale/fec.h b/drivers/net/ethernet/freescale/fec.h
index 33f84a30e1..5b4da21d8d 100644
--- a/drivers/net/ethernet/freescale/fec.h
+++ b/drivers/net/ethernet/freescale/fec.h
@@ -664,6 +664,36 @@ struct fec_enet_private {
 	u64 ethtool_stats[];
 };
 
+/*
+ * Check whether oob support is compiled in for the FEC driver. This
+ * tells nothing about the current execution stage, or whether oob
+ * diversion is ongoing for any FEC device.
+ */
+static inline bool fec_net_oob(void)
+{
+	return IS_ENABLED(CONFIG_FEC_OOB);
+}
+
+/*
+ * Check whether the caller is running on the out-of-band stage, with
+ * the precondition that oob support is compiled in for the FEC driver.
+ */
+static inline bool fec_running_oob(void)
+{
+	return fec_net_oob() && running_oob();
+}
+
+/*
+ * Check whether a FEC device is currently diverting packet to the
+ * out-of-band netstack. The compiler is given the opportunity to
+ * compile the whole thing out if either NET_OOB or FEC_OOB which
+ * depends on it are disabled.
+ */
+static inline bool fec_oob_enabled(struct fec_enet_private *fep)
+{
+	return IS_ENABLED(CONFIG_FEC_OOB) && netif_oob_diversion(fep->netdev);
+}
+
 void fec_ptp_init(struct platform_device *pdev, int irq_idx);
 void fec_ptp_stop(struct platform_device *pdev);
 void fec_ptp_start_cyclecounter(struct net_device *ndev);
diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index d10db5d6d2..9afcf58800 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -341,6 +341,50 @@ MODULE_PARM_DESC(macaddr, "FEC Ethernet MAC address");
 
 static int mii_cnt;
 
+#ifdef CONFIG_FEC_OOB
+
+static int fec_enet_get_irq_cnt(struct platform_device *pdev);
+
+static int fec_enable_oob(struct net_device *ndev)
+{
+	struct fec_enet_private *fep = netdev_priv(ndev);
+	int nr_irqs = fec_enet_get_irq_cnt(fep->pdev), n, ret = 0;
+
+	napi_disable(&fep->napi);
+	netif_tx_lock_bh(ndev);
+
+	for (n = 0; n < nr_irqs; n++) {
+		ret = irq_switch_oob(fep->irq[n], true);
+		if (ret) {
+			while (--n > 0)
+				irq_switch_oob(fep->irq[n], false);
+			break;
+		}
+	}
+
+	netif_tx_unlock_bh(ndev);
+	napi_enable(&fep->napi);
+
+	return ret;
+}
+
+static void fec_disable_oob(struct net_device *ndev)
+{
+	struct fec_enet_private *fep = netdev_priv(ndev);
+	int nr_irqs = fec_enet_get_irq_cnt(fep->pdev), n;
+
+	napi_disable(&fep->napi);
+	netif_tx_lock_bh(ndev);
+
+	for (n = 0; n < nr_irqs; n++)
+		irq_switch_oob(fep->irq[n], false);
+
+	netif_tx_unlock_bh(ndev);
+	napi_enable(&fep->napi);
+}
+
+#endif	/* !CONFIG_FEC_OOB */
+
 static struct bufdesc *fec_enet_get_nextdesc(struct bufdesc *bdp,
 					     struct bufdesc_prop *bd)
 {
@@ -455,6 +499,12 @@ fec_enet_create_page_pool(struct fec_enet_private *fep,
 	};
 	int err;
 
+	if (fec_net_oob()) {
+		pp_params.flags |= PP_FLAG_PAGE_OOB;
+		/* An oob pool can't grow, so plan for extra space. */
+		pp_params.pool_size *= 2;
+	}
+
 	rxq->page_pool = page_pool_create(&pp_params);
 	if (IS_ERR(rxq->page_pool)) {
 		err = PTR_ERR(rxq->page_pool);
@@ -481,6 +531,43 @@ fec_enet_create_page_pool(struct fec_enet_private *fep,
 	return err;
 }
 
+static dma_addr_t get_dma_mapping(struct sk_buff *skb,
+				struct device *dev, void *ptr,
+				size_t size, enum dma_data_direction dir)
+{
+	dma_addr_t addr;
+
+	if (!fec_net_oob() || !skb_has_oob_storage(skb))
+		return dma_map_single(dev, ptr, size, dir);
+
+	/*
+	 * An oob-managed storage is already mapped by the page pool
+	 * it belongs to. We only need to to let the device get at the
+	 * pre-mapped DMA area for the specified I/O direction.
+	 */
+	addr = skb_oob_storage_addr(skb);
+	dma_sync_single_for_device(dev, addr, size, dir);
+
+	return addr;
+}
+
+static void release_dma_mapping(struct sk_buff *skb,
+				struct device *dev, dma_addr_t addr, size_t size,
+				enum dma_data_direction dir)
+{
+	if (!fec_net_oob() || !skb || !skb_has_oob_storage(skb)) {
+		dma_unmap_single(dev, addr, size, dir);
+	} else {
+		/*
+		 * An oob-managed storage should not be unmapped, this
+		 * operation is handled when required by the page pool
+		 * it belongs to. We only need to synchronize the CPU
+		 * caches for the specified I/O direction.
+		 */
+		dma_sync_single_for_cpu(dev, addr, size, dir);
+	}
+}
+
 static struct bufdesc *
 fec_enet_txq_submit_frag_skb(struct fec_enet_priv_tx_q *txq,
 			     struct sk_buff *skb,
@@ -542,8 +629,8 @@ fec_enet_txq_submit_frag_skb(struct fec_enet_priv_tx_q *txq,
 				swap_buffer(bufaddr, frag_len);
 		}
 
-		addr = dma_map_single(&fep->pdev->dev, bufaddr, frag_len,
-				      DMA_TO_DEVICE);
+		addr = get_dma_mapping(skb, &fep->pdev->dev, bufaddr, frag_len,
+				DMA_TO_DEVICE);
 		if (dma_mapping_error(&fep->pdev->dev, addr)) {
 			if (net_ratelimit())
 				netdev_err(ndev, "Tx DMA memory map failed\n");
@@ -564,8 +651,8 @@ fec_enet_txq_submit_frag_skb(struct fec_enet_priv_tx_q *txq,
 	bdp = txq->bd.cur;
 	for (i = 0; i < frag; i++) {
 		bdp = fec_enet_get_nextdesc(bdp, &txq->bd);
-		dma_unmap_single(&fep->pdev->dev, fec32_to_cpu(bdp->cbd_bufaddr),
-				 fec16_to_cpu(bdp->cbd_datlen), DMA_TO_DEVICE);
+		release_dma_mapping(NULL, &fep->pdev->dev, fec32_to_cpu(bdp->cbd_bufaddr),
+				fec16_to_cpu(bdp->cbd_datlen), DMA_TO_DEVICE);
 	}
 	return ERR_PTR(-ENOMEM);
 }
@@ -619,7 +706,7 @@ static int fec_enet_txq_submit_skb(struct fec_enet_priv_tx_q *txq,
 	}
 
 	/* Push the data cache so the CPM does not get stale memory data. */
-	addr = dma_map_single(&fep->pdev->dev, bufaddr, buflen, DMA_TO_DEVICE);
+	addr = get_dma_mapping(skb, &fep->pdev->dev, bufaddr, buflen, DMA_TO_DEVICE);
 	if (dma_mapping_error(&fep->pdev->dev, addr)) {
 		dev_kfree_skb_any(skb);
 		if (net_ratelimit())
@@ -630,8 +717,8 @@ static int fec_enet_txq_submit_skb(struct fec_enet_priv_tx_q *txq,
 	if (nr_frags) {
 		last_bdp = fec_enet_txq_submit_frag_skb(txq, skb, ndev);
 		if (IS_ERR(last_bdp)) {
-			dma_unmap_single(&fep->pdev->dev, addr,
-					 buflen, DMA_TO_DEVICE);
+			release_dma_mapping(skb, &fep->pdev->dev, addr,
+					buflen, DMA_TO_DEVICE);
 			dev_kfree_skb_any(skb);
 			return NETDEV_TX_OK;
 		}
@@ -728,7 +815,7 @@ fec_enet_txq_put_data_tso(struct fec_enet_priv_tx_q *txq, struct sk_buff *skb,
 			swap_buffer(data, size);
 	}
 
-	addr = dma_map_single(&fep->pdev->dev, data, size, DMA_TO_DEVICE);
+	addr = get_dma_mapping(skb, &fep->pdev->dev, data, size, DMA_TO_DEVICE);
 	if (dma_mapping_error(&fep->pdev->dev, addr)) {
 		dev_kfree_skb_any(skb);
 		if (net_ratelimit())
@@ -789,7 +876,7 @@ fec_enet_txq_put_hdr_tso(struct fec_enet_priv_tx_q *txq,
 		if (fep->quirks & FEC_QUIRK_SWAP_FRAME)
 			swap_buffer(bufaddr, hdr_len);
 
-		dmabuf = dma_map_single(&fep->pdev->dev, bufaddr,
+		dmabuf = get_dma_mapping(skb, &fep->pdev->dev, bufaddr,
 					hdr_len, DMA_TO_DEVICE);
 		if (dma_mapping_error(&fep->pdev->dev, dmabuf)) {
 			dev_kfree_skb_any(skb);
@@ -937,12 +1024,23 @@ fec_enet_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	unsigned short queue;
 	struct fec_enet_priv_tx_q *txq;
 	struct netdev_queue *nq;
-	int ret;
+	int ret = 0;
 
 	queue = skb_get_queue_mapping(skb);
 	txq = fep->tx_queue[queue];
 	nq = netdev_get_tx_queue(ndev, queue);
 
+	/*
+	 * Lock out any sender running from the alternate execution
+	 * stage from other CPUs (i.e. oob vs in-band). Clearly,
+	 * in-band tasks should refrain from sending output through an
+	 * oob-enabled device when aiming at the lowest possible
+	 * latency for the oob players, but we still allow shared use
+	 * for flexibility though, which comes in handy when a single
+	 * NIC only is available to convey both kinds of traffic.
+	 */
+	netif_tx_lock_oob(nq);
+
 	if (skb_is_gso(skb))
 		ret = fec_enet_txq_submit_tso(txq, skb, ndev);
 	else
@@ -950,9 +1048,13 @@ fec_enet_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	if (ret)
 		return ret;
 
-	entries_free = fec_enet_get_free_txdesc_num(txq);
-	if (entries_free <= txq->tx_stop_threshold)
-		netif_tx_stop_queue(nq);
+	if (running_inband()) {
+		entries_free = fec_enet_get_free_txdesc_num(txq);
+		if (entries_free <= txq->tx_stop_threshold)
+			netif_tx_stop_queue(nq);
+	}
+
+	netif_tx_unlock_oob(nq);
 
 	return NETDEV_TX_OK;
 }
@@ -1001,10 +1103,11 @@ static void fec_enet_bd_init(struct net_device *dev)
 			bdp->cbd_sc = cpu_to_fec16(0);
 			if (bdp->cbd_bufaddr &&
 			    !IS_TSO_HEADER(txq, fec32_to_cpu(bdp->cbd_bufaddr)))
-				dma_unmap_single(&fep->pdev->dev,
-						 fec32_to_cpu(bdp->cbd_bufaddr),
-						 fec16_to_cpu(bdp->cbd_datlen),
-						 DMA_TO_DEVICE);
+				release_dma_mapping(txq->tx_skbuff[i],
+						&fep->pdev->dev,
+						fec32_to_cpu(bdp->cbd_bufaddr),
+						fec16_to_cpu(bdp->cbd_datlen),
+						DMA_TO_DEVICE);
 			if (txq->tx_skbuff[i]) {
 				dev_kfree_skb_any(txq->tx_skbuff[i]);
 				txq->tx_skbuff[i] = NULL;
@@ -1463,10 +1566,10 @@ fec_enet_tx_queue(struct net_device *ndev, u16 queue_id)
 		skb = txq->tx_skbuff[index];
 		txq->tx_skbuff[index] = NULL;
 		if (!IS_TSO_HEADER(txq, fec32_to_cpu(bdp->cbd_bufaddr)))
-			dma_unmap_single(&fep->pdev->dev,
-					 fec32_to_cpu(bdp->cbd_bufaddr),
-					 fec16_to_cpu(bdp->cbd_datlen),
-					 DMA_TO_DEVICE);
+			release_dma_mapping(skb, &fep->pdev->dev,
+					fec32_to_cpu(bdp->cbd_bufaddr),
+					fec16_to_cpu(bdp->cbd_datlen),
+					DMA_TO_DEVICE);
 		bdp->cbd_bufaddr = cpu_to_fec32(0);
 		if (!skb)
 			goto skb_done;
@@ -1558,7 +1661,7 @@ fec_enet_new_rxbdp(struct net_device *ndev, struct bufdesc *bdp, struct sk_buff
 	if (off)
 		skb_reserve(skb, fep->rx_align + 1 - off);
 
-	bdp->cbd_bufaddr = cpu_to_fec32(dma_map_single(&fep->pdev->dev, skb->data, FEC_ENET_RX_FRSIZE - fep->rx_align, DMA_FROM_DEVICE));
+	bdp->cbd_bufaddr = cpu_to_fec32(get_dma_mapping(skb, &fep->pdev->dev, skb->data, FEC_ENET_RX_FRSIZE - fep->rx_align, DMA_FROM_DEVICE));
 	if (dma_mapping_error(&fep->pdev->dev, fec32_to_cpu(bdp->cbd_bufaddr))) {
 		if (net_ratelimit())
 			netdev_err(ndev, "Rx DMA memory map failed\n");
@@ -3613,6 +3716,10 @@ static const struct net_device_ops fec_netdev_ops = {
 	.ndo_tx_timeout		= fec_timeout,
 	.ndo_set_mac_address	= fec_set_mac_address,
 	.ndo_eth_ioctl		= fec_enet_ioctl,
+#ifdef CONFIG_FEC_OOB
+	.ndo_enable_oob		= fec_enable_oob,
+	.ndo_disable_oob	= fec_disable_oob,
+#endif
 	.ndo_set_features	= fec_set_features,
 };
 
@@ -4161,6 +4268,11 @@ fec_probe(struct platform_device *pdev)
 	fep->rx_copybreak = COPYBREAK_DEFAULT;
 	INIT_WORK(&fep->tx_timeout_work, fec_enet_timeout_work);
 
+	if (IS_ENABLED(CONFIG_FEC_OOB)) {
+		ndev->priv_flags |= IFF_OOB_CAPABLE;
+		netdev_info(ndev, "FEC device is oob-capable\n");
+	}
+
 	pm_runtime_mark_last_busy(&pdev->dev);
 	pm_runtime_put_autosuspend(&pdev->dev);
 
-- 
2.30.2

